### PR TITLE
librbd: don't call refresh from mirror::GetInfoRequest state machine

### PR DESCRIPTION
**Pull request by trociny**
Fixes: https://tracker.ceph.com/issues/43589
Signed-off-by: Mykola Golub <mgolub@suse.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by trociny** [comment]
@dillaman The refresh has already been remove during optimization in #32434, but it probably makes sense to merge this as a separate PR before merging #32434 to ease backporting?

**Comment by dillaman** [comment]
Yes, that would be better. It’s hard for me to check right now, but do the API methods already refresh if required?

**Comment by trociny** [comment]
@dillaman The API methods do not refresh. I interpreted your "move the refresh to the API callers" that it is a responsibility of the caller to do refresh before calling the API methods. But it looks you meant to do this in `api::Mirror::image_get_info`? Ok, updated.

File: src/librbd/api/Mirror.cc [hunk 1]
```diff
@@ -651,7 +651,13 @@ void Mirror<I>::image_get_info(I *ictx, mirror_image_info_t *mirror_image_info,
   auto req = mirror::GetInfoRequest<I>::create(*ictx, &ctx->mirror_image,
                                                &ctx->promotion_state,
                                                ctx);
-  req->send();
+  if (ictx->state->is_refresh_required()) {
```
Comment (dillaman) [inline src/librbd/api/Mirror.cc]: Nit: just use `ictx->state->refresh_if_required()`

Comment (trociny) [reply src/librbd/api/Mirror.cc]: We don't have async version of `refresh_if_required`. Do you want me to add it?

Comment (dillaman) [reply src/librbd/api/Mirror.cc]: Sorry, didn't realize that this was an async API method.  In reality, if the refresh fails we should not advance. That would then imply that we need to delay the creation of `C_ImageGetInfo` until after the refresh succeeds (if a refresh is required).

Comment (trociny) [reply src/librbd/api/Mirror.cc]: Just to make sure I understand what you suggest. Should I implement the refresh retry loop until it succeeds? Wouldn't just checking for refresh error and returning it be enough? Why this image_get_info so special? I don't see we do something like this in other places.

And what if the error is fatal, e.g. due to the client is blacklisted or the image is removed? We may loop forever.

Comment (dillaman) [reply src/librbd/api/Mirror.cc]: Right now, sync API methods call `refresh_if_required` and return the error code if the refresh failed. Therefore, if `is_refresh_required` is `true` and then `refresh` fails, you should return that failure code to the async caller.

**Comment by trociny** [comment]
@dillaman updated

**Comment by dillaman** [comment]
jenkins test make check

**Pull request merged**